### PR TITLE
Support ConjugateTranspose for TF

### PR DESCRIPTION
**Pull request by Abdulrahman-Adel**
### Details:
* support ConjugateTranspose for TF
* Create src/frontends/tensorflow_common/src/op/conj_transpose.cpp
* Update src/frontends/tensorflow_common/include/common_op_table.hpp
* Update src/frontsends/tensorflow/src/op_table.cpp

### Tickets:
* Closes this issue [Support ConjugateTranspose operation for TensorFlow models](https://github.com/openvinotoolkit/openvino/issues/21463)

### CODE REVIEW

**Comment by popovaan** [comment]
Please, add a layer test for `ConjugateTranspose`.
Here is example: https://github.com/openvinotoolkit/openvino/blob/6060683f4c1ac320f1c01fc5c8fc8b8f53f93f02/tests/layer_tests/tensorflow_tests/test_tf_Cumsum.py#L13

File: tools/mo/unit_tests/moc_tf_fe/test_models/model_fp32.frozen
Comment (popovaan) [inline tools/mo/unit_tests/moc_tf_fe/test_models/model_fp32.frozen:1]: Is this change needed?

File: tools/ovc/unit_tests/moc_tf_fe/test_models/model_fp32.frozen
Comment (popovaan) [inline tools/ovc/unit_tests/moc_tf_fe/test_models/model_fp32.frozen:1]: Is this change needed?

**Comment by rkazants** [comment]
Hi @Abdulrahman-Adel,

Please address comments from @popovaan.

Best regards,
Roman

**Comment by Abdulrahman-Adel** [comment]
Hello @rkazants and @popovaan, 
thank you both for your comments. I've added a test layer and finished all necessary changes.

File: tests/layer_tests/tensorflow_tests/test_tf_ConjugateTranspose.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+        import tensorflow as tf
+
+        tf.compat.v1.reset_default_graph()
+
+        # Create the graph and model
+        with tf.compat.v1.Session() as sess:
+            tf_input = tf.compat.v1.placeholder(tf.complex64, shape, "Input")
+
+            tf.raw_ops.ConjugateTranspose(x=tf_input, perm=perm, name = "Operation")
```
Comment (rkazants) [inline tests/layer_tests/tensorflow_tests/test_tf_ConjugateTranspose.py:37]: not quite correct, please hide complex type to intermediate layers because we don't support complex type for Parameters and Result nodes. We just imitate this support by expressing a tensor of complex type as floating-point tensor with auxiliary dimension.
Example of such layer test for operation with complex type you can see here: https://github.com/openvinotoolkit/openvino/blob/master/tests/layer_tests/tensorflow_tests/test_tf_ComplexFFT.py.
You should pass real and imag parts from outside.

File: src/frontends/tensorflow_common/src/op/conj_transpose.cpp [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+	auto imag_index = make_shared<v0::Constant>(element::i32, Shape{}, 1);
+
+    auto gather_axis = make_shared<v0::Constant>(element::i32, Shape{1}, -1);
+	
+	auto real = make_shared<v8::Gather>(data, real_index, gather_axis)->output(0);
+    auto imag = make_shared<v8::Gather>(data, imag_index, gather_axis)->output(0);
+
+	
+	auto const_minus_one = make_shared<v0::Constant>(element::i32, Shape{}, -1);
+	imag = make_shared<v0::MatMul>(imag, const_minus_one, false, false);
```
Comment (rkazants) [inline src/frontends/tensorflow_common/src/op/conj_transpose.cpp:47]: please `Negative` operation instead of multiplication

File: src/frontends/tensorflow_common/src/op/conj_transpose.cpp [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+	
+
+	auto conj_tensor = make_shared<v0::Concat>(OutputVector{real, imag}, -1)->output(0);
+
+	auto conj_transpose = make_shared<v1::Transpose>(conj_tensor, perm);
```
Comment (rkazants) [inline src/frontends/tensorflow_common/src/op/conj_transpose.cpp:52]: pay attention to that `perm` has smaller length that `conj_tensor` due to presence of auxiliary dimension.
I propose to adjust this `perm`. For this you need to compute size of `perm` using `ShapeOf` operation and concatenate to this.

**Review by rkazants — COMMENTED** [review]
please follow code-style using clang

File: src/frontends/tensorflow_common/src/op/conj_transpose.cpp [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+		auto real = make_shared<v8::Gather>(x, real_index, gather_axis)->output(0);
+		auto imag = make_shared<v8::Gather>(x, imag_index, gather_axis)->output(0);
+
+		imag = make_shared<v0::Negative>(imag);
+	
+		auto conj_tensor = make_shared<v0::Concat>(OutputVector{real, imag}, -1)->output(0);
+
+		OutputVector concat_inputs;
+        concat_inputs.push_back(perm);
+        concat_inputs.push_back(make_shared<v0::Constant>(perm.get_element_type(), Shape{1}, 2));
```
Comment (rkazants) [inline src/frontends/tensorflow_common/src/op/conj_transpose.cpp:48]: I think this is not quite correct. Let us assume `perm = [1, 0, 2]` that means we have a deal with 3D tensor of complex type. So we should append `3` to perm to get `[1,0,2,3]` because we don't have to permute auxiliary dimension.

So I propose instead of this `Constant` to do the following (it may be preudo-code):

```
auto perm_size = make_shared<v0::ShapeOf>(perm, perm.get_element_type());
auto new_perm = auto concat = make_shared<v0::Concat>(OutputVector{perm, perm_size}, 0);
```

**Review by rkazants — COMMENTED** [review]
Left one comment. The other part looks good.
Also, please revert changes for `model_fp32.frozen` files

File: src/frontends/tensorflow_common/src/op/conj_transpose.cpp (hunk 4 above)
Comment (Abdulrahman-Adel) [reply src/frontends/tensorflow_common/src/op/conj_transpose.cpp]: Done @rkazants

File: src/frontends/tensorflow_common/src/op/conj_transpose.cpp [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+		auto real = make_shared<v8::Gather>(x, real_index, gather_axis)->output(0);
+		auto imag = make_shared<v8::Gather>(x, imag_index, gather_axis)->output(0);
+
+		imag = make_shared<v0::Negative>(imag);
+	
+		auto conj_tensor = make_shared<v0::Concat>(OutputVector{real, imag}, -1)->output(0);
+
+		OutputVector concat_inputs;
+        concat_inputs.push_back(perm);
+        concat_inputs.push_back(make_shared<v0::ShapeOf>(perm, perm.get_element_type()));
```
Comment (rkazants) [inline src/frontends/tensorflow_common/src/op/conj_transpose.cpp:49]:
```suggestion
        concat_inputs.push_back(make_shared<v3::ShapeOf>(perm, perm.get_element_type()));
```

**Review by rkazants — APPROVED** [review]

File: src/frontends/tensorflow_common/src/op/conj_transpose.cpp [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+		auto real_index = make_shared<v0::Constant>(element::i32, Shape{}, 0);
+		auto imag_index = make_shared<v0::Constant>(element::i32, Shape{}, 1);
+
+		auto gather_axis = make_shared<v0::Constant>(element::i32, Shape{1}, -1);
+		
+		auto real = make_shared<v8::Gather>(x, real_index, gather_axis)->output(0);
+		auto imag = make_shared<v8::Gather>(x, imag_index, gather_axis)->output(0);
+
+		imag = make_shared<v0::Negative>(imag);
+
```
Comment (rkazants) [inline src/frontends/tensorflow_common/src/op/conj_transpose.cpp:44]: please fix code-style according to clang format.

**Comment by rkazants** [comment]
Hi @Abdulrahman-Adel,

Please fix errors in new tests:

```
_ TestComplexConjugateTranspose.test_conjugate_transpose[ ie_device:CPU - precision:FP16 - params:{'shape': [1, 2], 'perm': [1, 0]} ] _

self = <test_tf_ConjugateTranspose.TestComplexConjugateTranspose object at 0x7f5bc2cd1110>
params = {'perm': [1, 0], 'shape': [1, 2]}, ie_device = 'CPU'
precision = 'FP16', ir_version = 11
temp_dir = '/__w/openvino/openvino/install/tests/layer_tests/common/out/CPU_test_conjugate_transpose6e2yr8go'
use_new_frontend = True, use_old_api = False

    @pytest.mark.parametrize("params", test_data)
    @pytest.mark.precommit_tf_fe
    @pytest.mark.nightly
    def test_conjugate_transpose(self, params, ie_device, precision, ir_version, temp_dir,
                                 use_new_frontend, use_old_api):
>       self._test(*self.create_complex_conjugate_transpose_net(**params),
                   ie_device, precision, ir_version, temp_dir=temp_dir,
                   use_new_frontend=use_new_frontend, use_old_api=use_old_api)
E       TypeError: TestComplexConjugateTranspose.create_complex_conjugate_transpose_net() got an unexpected keyword argument 'shape'
```

Logs are here: https://github.com/openvinotoolkit/openvino/actions/runs/7346453103/job/20001617537?pr=21586

BTW, maybe you have a problem with running layer tests. Here are cmd-lines of example:

```
# install pre-reqs from openvino/tests/layer_tests/requirements.txt
export TEST_DEVICE=CPU
pytest tensorflow_tests/test_tf_AddN.py -m precommit_tf_fe --use_new_frontend
```

File: src/frontends/tensorflow_common/src/op/conj_transpose.cpp (hunk 6 above)
Comment (Abdulrahman-Adel) [reply src/frontends/tensorflow_common/src/op/conj_transpose.cpp]: Hello @rkazants, I fixed it.

**Comment by rkazants** [comment]
build_jenkins

File: src/frontends/tensorflow_common/src/op/conj_transpose.cpp [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+using namespace std;
+using namespace ov::op;
+
+namespace ov {
+namespace frontend {
+namespace tensorflow {
+namespace op {
+
+OutputVector translate_conj_transpose_op(const NodeContext& node) {
+    default_op_checks(node, 2, {"ConjugateTranspose"});
```
Comment (rkazants) [inline src/frontends/tensorflow_common/src/op/conj_transpose.cpp:23]:
```suggestion
    default_op_checks(node, 2, {"ConjugateTranspose"}, true);
```

**Comment by rkazants** [comment]
build_jenkins

**Comment by Abdulrahman-Adel** [comment]
> build_jenkins

@rkazants please build again.

thank you for your patience

**Comment by rkazants** [comment]
build_jenkins

**Comment by rkazants** [comment]
> > build_jenkins
> 
> @rkazants please build again.
> 
> thank you for your patience

No problem, thanks for your effort!

File: src/frontends/tensorflow_common/src/op/conj_transpose.cpp [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+    auto x = node.get_input(0);
+    auto perm = node.get_input(1);
+
+    auto complex_type_mark = as_type_ptr<ComplexTypeMark>(x.get_node_shared_ptr());
+    if (complex_type_mark) {
+        element::Type complex_part_type = complex_type_mark->get_complex_part_type();
+        auto x = complex_type_mark->input_value(0);
+
+        auto real_index = make_shared<v0::Constant>(element::i32, Shape{}, 0);
+        auto imag_index = make_shared<v0::Constant>(element::i32, Shape{}, 1);
```
Comment (rkazants) [inline src/frontends/tensorflow_common/src/op/conj_transpose.cpp:34]:
```suggestion
        auto real_index = make_shared<v0::Constant>(element::i32, Shape{1}, 0);
        auto imag_index = make_shared<v0::Constant>(element::i32, Shape{1}, 1);
```

Comment (rkazants) [reply src/frontends/tensorflow_common/src/op/conj_transpose.cpp:34]: that auxiliary dimension should be preserved. otherwise, we seem to loose it and rank will be lower for `Transpose` operation below

**Comment by rkazants** [comment]
build_jenkins
